### PR TITLE
uArch dependency support

### PR DESCRIPTION
**Pull request by iamkarthikbk**
WARL CSRs can now include hidden uArch dependencies in their YAML definitions.

the `warl_class` now creates a `uarch_depends` dict to support adding hidden uArch
signals to the dependency string in a CSR's YAML definition.

1. The prefix for such a signal/group must be `uarch_`. For example -
  - `uarch_cachecontrol` is a valid group/signal. In order to group signals under this
    name, `uarch_cachecontrol::global_valid` and `uarch_cachecontrol::global_dirty` are
    valid dependencies.
  - `uarch_global_valid` and `uarch_global_dirty` are valid signals. Such signals, for
    parsing purposes are grouped with an implicit name `uarch_commonconfig`. This name
    can not be used for any other micro-architectural signal dependency.

2. Any such signals/groups will be excluded from error checks, and will evaluate to
   `True` for legal checks.

3. Upon encountering such a signal/group, a warning statement will be logged:

```text
[DEBUG]   riscv_config.checker: -- Checking reset values for csr: mtvec::base
[WARNING] riscv_config.warl: WARL for csr mtvec::base depends on uarch csr uarch_global_valid. Treating this as a uarch dependency and excluding from all furtherchecks.
[DEBUG]   riscv_config.warl: uArch dependencies are: {'uarch_commonconfig':['uarch_global_valid']}
[WARNING] riscv_config.warl: WARL for csr mtvec::base depends on uarch csr uarch_global_dirty. Treating this as a uarch dependency and excluding from all further checks.
[DEBUG]   riscv_config.warl: uArch dependencies are: {'uarch_commonconfig': ['uarch_global_valid', 'uarch_global_dirty']}
[DEBUG]   riscv_config.warl: ---- WARL Value Legality Check: value:0 csr:mtvec::base dependency_vals:None
[WARNING] riscv_config.warl: csr/subfield uarch_global_valid is a uArch dependency.
[WARNING] riscv_config.warl: csr/subfield uarch_global_dirty is a uArch dependency.
[DEBUG]   riscv_config.warl:  warl legal string "uarch_global_valid[3:0] in [0] and uarch_global_dirty[3:0] in [0] -> base[61:0] bitmask [0x3FFFFFFFFFFFFFFF, 0x0000000000000000]" treats the input value 0 as legal
```

4. Such signals will show up in the checked YAMLs as follows:

if grouped --
```text
warl:
  dependency_fields: [uarch_cachecontrol::global_valid, uarch_cachecontrol::global_dirty]
  legal:
    - global_valid[3:0] in [0] and global_dirty[3:0] in [0]
      -> base[61:0] bitmask [0x3FFFFFFFFFFFFFFF, 0x0000000000000000]
  wr_illegal:
    - Unchanged
```

or if not grouped --
```text
warl:
    dependency_fields: [uarch_global_valid, uarch_global_dirty]
    legal:
        - uarch_global_valid[3:0] in [0] and uarch_global_dirty[3:0]
        in [0] -> base[61:0] bitmask [0x3FFFFFFFFFFFFFFF, 0x0000000000000000]
    wr_illegal:
        - Unchanged
```

### CODE REVIEW

**Comment by pawks** [comment]
Please update the docs for warl([link](https://github.com/riscv-software-src/riscv-config/blob/master/docs/source/yaml-specs.rst#warl-node-definition)) with the corresponding information.

**Comment by iamKarthikBK** [comment]
ready for review @pawks

File: riscv_config/warl.py [hunk 1]
```diff
@@ -55,9 +56,12 @@ def __init__(self, node, csrname, f_msb, f_lsb, spec=None):
         self.regex_legal = re.compile('\[(.*?)\]\s*(bitmask|in|not in)\s*\[(.*?)\]')
         self.regex_dep_legal = re.compile('(?P<csrname>.*?)\s*\[(?P<indices>.*?)\]\s*(?P<op>in|not in.*?)\s*\[(?P<values>.*?)\]\s*')
         self.dependencies = node['dependency_fields']
+        self.uarch_depends = {'uarch_signals': []}
+        self.uarch_signals = uarch_signals
```
Comment (pawks) [inline riscv_config/warl.py:60]: Can be derived from the `spec` variable itself. Prevents making a non-backward compatible API change.

File: riscv_config/schemas/schema_custom.yaml [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+              lsb:
+                type: integer
+                required: true
+                min: 1
+                max: 63
+              reset-val:
+                type: integer
+                required: true
+                min: 0
+                max: 0xFFFFFFFFFFFFFFFF
```
Comment (pawks) [inline riscv_config/schemas/schema_custom.yaml:154]: No Max value required.

File: riscv_config/schemas/schema_custom.yaml [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+                  type: integer
+          - schema:
+              subfields:
+                type: dict
+                schema:
+                  msb:
+                    type: integer
+                    required: true
+                    min: 0
+                    max: 62
```
Comment (pawks) [inline riscv_config/schemas/schema_custom.yaml:169]: No max value required.

File: riscv_config/schemas/schema_custom.yaml [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+                  lsb:
+                    type: integer
+                    required: true
+                    min: 1
+                    max: 63
```
Comment (pawks) [inline riscv_config/schemas/schema_custom.yaml:174]: No Max value required.

File: riscv_config/schemas/schema_custom.yaml [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                schema:
+                  msb:
+                    type: integer
+                    required: true
+                    min: 0
+                    max: 62
+                  lsb:
+                    type: integer
+                    required: true
+                    min: 1
```
Comment (pawks) [inline riscv_config/schemas/schema_custom.yaml:173]: min of lsb cant be greater than msb. Additionally, the way schema is defined, each subfield will have to be a minimum of 2 bits wide. No min-max bounds are necessary for these registers. The schema would also benefit from an additional check_with function to ensure that the subfields are non-overlapping.

**Review by pawks — CHANGES_REQUESTED** [review]

**Review by pawks — APPROVED** [review]
LGTM!

**Pull request merged**